Fixed rendered template values not displaying with proper code format…

dheerajturaga opened the pull request:
This PR attempts to fix #50324 , #49064

  The issue was caused by double JSON serialization in the frontend:
  1. Backend already serializes template field values to strings via serialize_template_field()
  2. Frontend was calling JSON.stringify() again on these already-stringified values
  3. This resulted in escaped quotes and poor formatting

<img width="1438" height="675" alt="image" src="https://github.com/user-attachments/assets/a290069f-6f15-45c2-bb0f-76654ca6b60c" />


Core Fixes

  - Fixed double serialization: Replaced JSON.stringify(value) with proper type checking in RenderedTemplates.tsx:97
    - Use string values directly if already strings
    - Only JSON stringify non-string values (objects/arrays)
  - Added whitespace preservation: Replaced \<Code\> component with \<SyntaxHighlighter\> to preserve newlines and indentation

Syntax Highlighting Enhancement

  - Integrated react-syntax-highlighter with automatic language detection:
    - JSON: Detects objects/arrays with valid JSON structure
    - YAML: Detects key-value pairs with colons and lists
    - SQL: Detects keywords like SELECT, INSERT, CREATE, etc.
    - Bash: Detects commands, shebangs, control structures, and operators
    - Text: Falls back to plain text for unrecognized content
  - Theme support: Respects dark/light mode themes consistent with DAG code viewer
  - Improved styling: Added proper spacing, border radius, and responsive alignment

dheerajturaga committed:
Fixed rendered template values not displaying with proper code formatting and syntax highlighting in the Task Instance UI

dheerajturaga committed:
Add line numbers

jscheffl commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/RenderedTemplates.tsx:
```diff
@@ -16,14 +16,75 @@
+SyntaxHighlighter.registerLanguage("sql", sql);
+SyntaxHighlighter.registerLanguage("bash", bash);
+
+const detectLanguage = (value: string): string => {
```
Can you separate this function out into a separate TS module file?

dheerajturaga replied:
Done!

jscheffl commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/RenderedTemplates.tsx:
```diff
@@ -16,14 +16,75 @@
+  }
+
+  // Try to detect YAML (basic heuristics)
+  if (trimmed.includes(":") && (trimmed.includes("\n") || trimmed.includes("- "))) {
```
Detection safety:
- If a colon is matched it should be followed by a newline (`":\n"`)
- Should we not rather use a lib like https://www.npmjs.com/package/yaml to test parsing after some basic heuristics before making a decision (similar to JSON)?

dheerajturaga replied:
Done!

jscheffl commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/RenderedTemplates.tsx:
```diff
@@ -16,14 +16,75 @@
+  // Try to detect SQL (basic heuristics)
+  const sqlKeywords = /\b(?:select|insert|update|delete|create|alter|drop|from|where|join)\b/iu;
+
+  if (sqlKeywords.test(trimmed)) {
```
Would it might be better to match more than a single keyword, as `SELECT` will in almost all cases will be followed by a `FROM` and otherwise `DROP` will also usually be followed by a `TABLE`.

dheerajturaga replied:
I switched to using a SQL parser. should be more robust!

jscheffl reviewed (commented):
From reading code I really like the lean approach. Some matching nit from my side to make it a bit more safe.

Would like to have an UI expert second opinion though.

dheerajturaga committed:
Seperate out detectLanguage into its own ts module

dheerajturaga committed:
Added actual YAML parsing for validation

dheerajturaga committed:
Added actual SQL parsing for validation

jscheffl reviewed (approved):
For me this looks good! LGTM!

pierrejeambrun commented on airflow-core/src/airflow/ui/src/pages/TaskInstance/RenderedTemplates.tsx:
```diff
@@ -32,20 +46,42 @@ export const RenderedTemplates = () => {
+                            padding: "8px",
+                          },
+                        }}
+                        flex="1"
```
Can we not use such inline css, and style the appropriate component. Also chakraUI provide some default value for padding marging etc, that are 1,2,3,4,5 and translate to fix values in pixels for consitency accross the UI.

dheerajturaga replied:
@pierrejeambrun , Done! let me know if this is good!

pierrejeambrun commented on airflow-core/src/airflow/ui/src/utils/detectLanguage.ts:
```diff
@@ -0,0 +1,90 @@
+  if (
+    (trimmed.startsWith("{") && trimmed.endsWith("}")) ||
+    (trimmed.startsWith("[") && trimmed.endsWith("]"))
+  ) {
```
I would remove that check, Just try parsing it if it's not valid, then it's not json

jscheffl replied:
I think this is good to be kept. It is a clever performance optimization because this is a good pre-check before a "heavy" parser gets going... if it is a longer text body then this saves some time and makes the page faster I think.

dheerajturaga replied:
yea, I agree. The pre check avoids parsing which could be helpful if these fields are large. I cleanup if you think otherwise.

pierrejeambrun replied:
I'm pretty sure that this is exacly what all parsers will do. 

For instance `rhino` js engine (mozilla js engine written in Java), you can find that:
https://github.com/mozilla/rhino/blob/master/rhino/src/main/java/org/mozilla/javascript/json/JsonParser.java#L53

`JSON.parse` will fail early if characters are not JSON parsable.

dheerajturaga replied:
@pierrejeambrun , point taken. made the changes you have requested. Please let me know if these are good

jscheffl replied:
Okay, hope the same is with YAML... then this is a convincing argument...

pierrejeambrun commented on airflow-core/src/airflow/ui/src/utils/detectLanguage.ts:
```diff
@@ -0,0 +1,90 @@
+    }
+  } catch {
+    // Not valid YAML, continue to other checks
+  }
```
Same here

dheerajturaga replied:
Done!

pierrejeambrun reviewed (approved):
Just a few minor comments, looks good to me.

dheerajturaga committed:
Replace inline CSS with Chakra UI design tokens in RenderedTemplates

dheerajturaga committed:
Remove string-based heuristics from JSON/YAML detection in detectLanguage

Rely solely on parser exceptions rather than basic string pattern matching.
The JSON.parse() and parseYaml() functions already handle validation
internally, making the manual checks for brackets and colons redundant.

dheerajturaga committed:
Improve language detection order and add debug badges

Reorder detectLanguage priority: JSON → SQL → Bash → YAML → Text.
The YAML parser is overly permissive and was incorrectly detecting
bash scripts as YAML. Moving bash detection before YAML fixes this.

jscheffl reviewed (approved):
Good for me!
